### PR TITLE
Implement #212 - Increasing distance stop times verification

### PR DESCRIPTION
**Pull request by lionel-nj**
**Summary:**

This PR provides support to verify that for each stop time of a trip have increasing `shape_dist_travelled` valued

**Expected behavior:** 

The Python validator code treats this rule as follows:
- [x] two consecutive stops times have same value for `shape_dist_travelled`-> this is an error
- [x] two consecutive stop times have values of `shape_dist_travelled` not ordered by ascending order -> this is an error

Code fro the Python valdiator https://github.com/MobilityData/transitfeed/blob/54a4081b59bfa015d5f0405b68203e61762d4a52/transitfeed/trip.py#L594

Tests from the python validator

https://github.com/MobilityData/transitfeed/blob/54a4081b59bfa015d5f0405b68203e61762d4a52/tests/transitfeed/testtrip.py#L357

Explain and/or show screenshots for how you expect the pull request to work in your testing (in case other devices exhibit different behavior).

Capture d’écran, le 2020-08-25 à 11 29 47

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "Fix #<issue_number> - <short description of fix and changes>" (for example - "Fix #1111 - Check for null value before using field")
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java [hunk 1]
```diff
@@ -854,6 +855,36 @@ public void export(final UnusableTripNotice toExport) throws IOException {
...
+    }
+
+    @Override
+    public void export(final DecreasingStopTimeDistanceWarningNotice toExport) throws IOException {
+        decreasingStopTimeDistanceToProto(GtfsValidationOutputProto.GtfsProblem.Severity.WARNING, toExport);
+    }
+
+    private void decreasingStopTimeDistanceToProto(final GtfsValidationOutputProto.GtfsProblem.Severity severity,
+                                                   final Notice toExport)
+            throws IOException {
```
Comment (barbeau) [inline adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java:871]: Could be moved to above line

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/DecreasingStopTimeDistanceErrorNotice.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+public class DecreasingStopTimeDistanceErrorNotice extends ErrorNotice {
+    public DecreasingStopTimeDistanceErrorNotice(final String tripId,
+                                                 final int stopSequence,
+                                                 final Float shapeDistTraveled,
+                                                 final int conflictingStopSequence,
+                                                 final Float conflictingShapeDistTraveled) {
+        super("stop_times.txt", E_054,
+                "Decreasing `shape_dist_traveled` values",
+                String.format("Trip with `trip_id` `%s` the stop `%s` has shape_dist_traveled=`%s`, which should be" +
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/DecreasingStopTimeDistanceErrorNotice.java:32]: I'd suggest a more succinct message that includes the values close the front of the message, so you don't have to scroll to see them, as well as the exact field names:

>`trip_id` 1234 `stop_sequence` 1 has a larger `shape_dist_traveled` (15.21) than `stop_sequence` 2 (12.35). `shape_dist_traveled` must increase with `stop_sequence`.

File: usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateStopTimeIncreasingDistance.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    public void execute(final TreeMap<Integer, StopTime> stopTimeSequence) {
+        // get previous stoptime relevant information
+        final var previousStopTimeData = new Object() {
+            Float shapeDistTraveled = null;
+            Integer stopSequence = null;
+        };
+        stopTimeSequence.forEach((stopSequence, stopTime) -> {
+            final Float shapeDistTraveled = stopTime.getShapeDistTraveled();
+            if (shapeDistTraveled != null && previousStopTimeData.shapeDistTraveled != null) {
+                if (previousStopTimeData.shapeDistTraveled > shapeDistTraveled) {
```
Comment (barbeau) [inline usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateStopTimeIncreasingDistance.java:55]: See my above comment - I believe this should be `>=`

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/DecreasingStopTimeDistanceErrorNotice.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                                " larger than the previous ones. In this case, the previous distance was `%s` (for the " +
+                                "stop `%s`).",
+                        tripId, conflictingStopSequence, conflictingShapeDistTraveled, shapeDistTraveled, stopSequence),
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/DecreasingStopTimeDistanceErrorNotice.java:35]: I would try to keep the order of variables here and in the constructor the same - it could be easy to accidentally swap them.

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/DecreasingStopTimeDistanceErrorNotice.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+package org.mobilitydata.gtfsvalidator.domain.entity.notice.error;
+
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.NoticeExporter;
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.base.ErrorNotice;
+
+import java.io.IOException;
+
+public class DecreasingStopTimeDistanceErrorNotice extends ErrorNotice {
+    public DecreasingStopTimeDistanceErrorNotice(final String tripId,
+                                                 final int stopSequence,
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/DecreasingStopTimeDistanceErrorNotice.java:26]: I would rename these variables to define the order of sequence, like:
* `stopSequence`
* `shapeDistTraveled`
* `previousStopSequence`
* `previousShapeDistTraveled`

Per my comment below, without explicit order in the names it would be easy to accidentally swap the variables when calling the method, unless you looked at the implementation.

File: usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateStopTimeIncreasingDistanceTest.java [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+import java.util.TreeMap;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mobilitydata.gtfsvalidator.domain.entity.notice.base.Notice.*;
+import static org.mockito.Mockito.*;
+
+class ValidateStopTimeIncreasingDistanceTest {
+
+    // suppressed warning regarding ignored result of method used in assertions
+    @SuppressWarnings("ResultOfMethodCallIgnored")
```
Comment (barbeau) [inline usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateStopTimeIncreasingDistanceTest.java:35]: I would add one more test where the first stop sequence shape_dist_traveled is 10 and the second 5, just to cover all corner cases (i.e., make sure a large value in the first sequence index is detected).

File: adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java (hunk 1 above)
Comment (lionel-nj) [reply adapter/exporter/src/main/java/org/mobilitydata/gtfsvalidator/exporter/ProtobufNoticeExporter.java:871]: Since the warning notice has been removed, this method is no more used.

File: RULES.md [hunk 7]
```diff
@@ -307,6 +309,12 @@ Trips must be referred to at least once in `stop_times.txt`.
 
 A trip must visit more than one stop in `stop_times.txt` to be usable by passengers for boarding and alighting.
 
+<a name="E054"/>
+
+### E054 - Decreasing travelled distance
```
[the remark this answers is not in the document]
Comment (lionel-nj) [reply RULES.md:314]: Will implement this rule in another PR

[the remark this answers is not in the document]
Comment (lionel-nj) [reply RULES.md]: See https://github.com/MobilityData/gtfs-validator/issues/358
